File: src/tqi/scoring/amenity.py
"""Amenity accessibility: can people reach essential destinations by transit?"""

import json
from pathlib import Path

import numpy as np

from tqi.config import DATA_DIR, WALK_SPEED_KMH

AMENITIES_FILE = DATA_DIR / "amenities.json"

INF = float("inf")

# Time thresholds for reporting (minutes)
TIME_THRESHOLDS = [30, 45, 60]


def load_amenities(path: Path = AMENITIES_FILE) -> list[dict]:
    """Load amenity definitions from JSON."""
    if not path.exists():
        return []
    with open(path) as f:
        return json.load(f)


def compute_amenity_accessibility(
    grid: np.ndarray,
    mean_travel_time: np.ndarray,
    distances_km: np.ndarray,
    amenities: list[dict],
) -> list[dict]:
    """For each amenity, find the nearest grid point and compute accessibility.

    Returns list of dicts with amenity info + access stats.
    """
    from haversine import haversine, Unit

    n_grid = len(grid)
    results = []

    for amenity in amenities:
        a_lat, a_lon = amenity["lat"], amenity["lon"]

        # Find the closest grid point to this amenity
        dists_to_amenity = np.array([
            haversine((a_lat, a_lon), (grid[i, 0], grid[i, 1]), unit=Unit.KILOMETERS)
            for i in range(n_grid)
        ])
        amenity_grid_idx = int(np.argmin(dists_to_amenity))

        # Travel times from all grid points TO this amenity's grid point
        # mean_travel_time[origin, dest] — we want column amenity_grid_idx
        tt_to_amenity = mean_travel_time[:, amenity_grid_idx]

        # Walking competitor: exclude if walking is faster
        walk_times = distances_km[:, amenity_grid_idx] / (WALK_SPEED_KMH / 60.0)
        effective_tt = np.where(
            (tt_to_amenity < INF) & (tt_to_amenity < walk_times),
            tt_to_amenity,
            INF,
        )

        # Compute % reachable within each time threshold
        threshold_pcts = {}
        for t in TIME_THRESHOLDS:
            reachable = (effective_tt <= t).sum()
            threshold_pcts[t] = float(reachable / n_grid * 100)

        any_reachable = (effective_tt < INF).sum()
        median_tt = float(np.median(effective_tt[effective_tt < INF])) if any_reachable > 0 else None

        results.append({
            "name": amenity["name"],
            "category": amenity["category"],
            "lat": a_lat,
            "lon": a_lon,
            "grid_idx": amenity_grid_idx,
            "pct_within_30min": threshold_pcts.get(30, 0),
            "pct_within_45min": threshold_pcts.get(45, 0),
            "pct_within_60min": threshold_pcts.get(60, 0),
            "pct_reachable": float(any_reachable / n_grid * 100),
            "median_travel_time": median_tt,
        })

    return results

File: go/internal/scoring/amenity.go
package scoring

import (
	"encoding/json"
	"math"
	"os"

	"github.com/jasondyck/chwk-tqi/internal/geo"
	"github.com/jasondyck/chwk-tqi/internal/grid"
)

// Amenity represents a point of interest.
type Amenity struct {
	Name     string `json:"name"`
	Lat      float64 `json:"lat"`
	Lon      float64 `json:"lon"`
	Category string `json:"category"`
}

// AmenityResult holds accessibility metrics for a single amenity.
type AmenityResult struct {
	Name            string  `json:"name"`
	Category        string  `json:"category"`
	Lat             float64 `json:"lat"`
	Lon             float64 `json:"lon"`
	PctWithin30Min  float64 `json:"pct_within_30_min"`
	PctWithin60Min  float64 `json:"pct_within_60_min"`
	MeanTravelTime  float64 `json:"mean_travel_time"`
}

// LoadAmenities reads amenities from a JSON file.
func LoadAmenities(path string) ([]Amenity, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var amenities []Amenity
	if err := json.Unmarshal(data, &amenities); err != nil {
		return nil, err
	}
	return amenities, nil
}

// ComputeAmenityAccessibility calculates how accessible each amenity is from
// the grid points, using the OD travel time matrix.
func ComputeAmenityAccessibility(points []grid.Point, meanTT, distKM [][]float64, amenities []Amenity) []AmenityResult {
	n := len(points)
	results := make([]AmenityResult, len(amenities))

	for a, am := range amenities {
		// Find nearest grid point to amenity.
		bestIdx := -1
		bestDist := math.MaxFloat64
		for j := 0; j < n; j++ {
			d := geo.Haversine(am.Lat, am.Lon, points[j].Lat, points[j].Lon)
			if d < bestDist {
				bestDist = d
				bestIdx = j
			}
		}

		results[a] = AmenityResult{
			Name:     am.Name,
			Category: am.Category,
			Lat:      am.Lat,
			Lon:      am.Lon,
		}

		if bestIdx < 0 || bestIdx >= len(meanTT) {
			continue
		}

		var within30, within60 int
		var ttSum float64
		var ttCount int

		for i := 0; i < n; i++ {
			if i >= len(meanTT) || bestIdx >= len(meanTT[i]) {
				continue
			}
			tt := meanTT[i][bestIdx]
			if math.IsInf(tt, 1) || tt >= InfFloat {
				continue
			}
			ttSum += tt
			ttCount++
			if tt <= 30 {
				within30++
			}
			if tt <= 60 {
				within60++
			}
		}

		if n > 0 {
			results[a].PctWithin30Min = float64(within30) / float64(n) * 100.0
			results[a].PctWithin60Min = float64(within60) / float64(n) * 100.0
		}
		if ttCount > 0 {
			results[a].MeanTravelTime = ttSum / float64(ttCount)
		}
	}

	return results
}
